packages/core/src/components/axes/ruler.ts

```diff
@@ -0,0 +1,188 @@
+		const rangeScale = this.services.cartesianScales.getRangeScale();
+		const [yScaleEnd, yScaleStart] = rangeScale.range();
+
+		const scaledData: number[] = displayData.map((d, i) =>
+			this.services.cartesianScales.getDomainValue(d, i)
+		);
+
+		/**
+		 * Find matches, reduce is used instead of filter
+		 * to only get elements which belong to the same axis coordinate
+		 */
+		const dataPointsMatchingRulerLine: number[] = scaledData.reduce((accum, currentValue) => {
+			// store the first element of the accumulator array to compare it with current element being processed
+			const sampleAccumValue = accum[0];
+
+			// if accumulator is not empty and current value is bigger than already existing value in the accumulator, skip current iteration
+			if (sampleAccumValue !== undefined && currentValue > sampleAccumValue) {
+				return accum;
+			}
+
+			// there's a match and currentValue is either less then or equal to already stored values
+			if (pointIsWithinThreshold(currentValue, x)) {
+				if (sampleAccumValue !== undefined && currentValue < sampleAccumValue) {
+					// there's a closer data point in the threshold area, so reinstantiate array
+					accum = [currentValue];
+				} else {
+					// currentValue is equal to already stored values, there's another match on the same coordinate
+					accum.push(currentValue);
+				}
+			}
+
+			return accum;
+		}, []);
+
+		// some data point match
+		if (dataPointsMatchingRulerLine.length > 0) {
+			const sampleMatch = dataPointsMatchingRulerLine[0];
+
+			const highlightItems = this.services.cartesianScales.getDataFromDomain(
+				domainScale.invert(sampleMatch)
+			).filter(d => d.value);
+
+			// get elements on which we should trigger mouse events  
+			const hoveredElements = dataPointElements.filter((d, i) =>
+				dataPointsMatchingRulerLine.includes(
+					Number(this.services.cartesianScales.getDomainValue(d))
+				)
+			);
+
+			/** if we pass from a trigger area to another one
+			 * mouseout on previous elements won't get dispatched
+			 * so we need to do it manually
+			 */
+			if (
+				this.hoveredElements &&
+				this.hoveredElements.size() > 0 &&
+				!Tools.isEqual(this.hoveredElements, hoveredElements)
+			) {
+				this.hideRuler();
+			}
+
+			hoveredElements.dispatch("mouseover");
+
+			// set current hovered elements
+			this.hoveredElements = hoveredElements;
+
+			this.services.events.dispatchEvent("show-tooltip", {
+				hoveredElement: line,
+				multidata: highlightItems,
+				type: TooltipTypes.GRIDLINE
+			});
+
+			ruler.attr("opacity", 1);
+
+			// line snaps to matching point
+			line.attr("y1", yScaleStart)
+				.attr("y2", yScaleEnd)
+				.attr("x1", sampleMatch)
+				.attr("x2", sampleMatch);
+		} else {
+			ruler.attr("opacity", 0);
+			dataPointElements.dispatch("mouseout");
+		}
+	}
+
+	hideRuler() {
+		const svg = this.parent;
+		const ruler = DOMUtils.appendOrSelect(svg, "g.ruler");
+		const dataPointElements = svg.selectAll("[role=graphics-symbol]");
```

Comment:
why use role to select instead of something like  `svg.selectAll("circle.dot");` ? 
if it also works, it seems more conventional in this codebase

Reply:
The use of that selector seemed to me more appropriate since we initially thought to enable the ruler on very different charts, which might have different graphic elements that need to be highlighted (for example, if I remember correctly there was an idea to have other shapes for the scatterplot other than circles - squares, triangles etc) that way we could have a single selector which includes all different svg elements. 
But you're right, at the moment the `circle.dot` selector should work across all current scenarios 👍